web/app.js

```diff
@@ -2743,42 +2783,94 @@ function webViewerTouchMove(evt) {
   }
 
   const { pdfViewer, _touchInfo, supportsPinchToZoom } = PDFViewerApplication;
-  const [touch0, touch1] = evt.touches;
+  let [touch0, touch1] = evt.touches;
+  if (touch0.identifier > touch1.identifier) {
+    [touch0, touch1] = [touch1, touch0];
+  }
   const { pageX: page0X, pageY: page0Y } = touch0;
   const { pageX: page1X, pageY: page1Y } = touch1;
-  const distance = Math.hypot(page0X - page1X, page0Y - page1Y) || 1;
-  const { distance: previousDistance } = _touchInfo;
-  const scaleFactor = distance / previousDistance;
+  const {
+    touch0X: pTouch0X,
+    touch0Y: pTouch0Y,
+    touch1X: pTouch1X,
+    touch1Y: pTouch1Y,
+  } = _touchInfo;
 
-  if (supportsPinchToZoom && Math.abs(scaleFactor - 1) <= 1e-2) {
-    // Scale increase/decrease isn't significant enough.
+  if (
+    Math.abs(pTouch0X - page0X) <= 1 &&
+    Math.abs(pTouch0Y - page0Y) <= 1 &&
+    Math.abs(pTouch1X - page1X) <= 1 &&
+    Math.abs(pTouch1Y - page1Y) <= 1
+  ) {
+    // Touches are really too close and it's hard do some basic
+    // geometry in order to guess something.
     return;
   }
 
-  const { centerX, centerY } = _touchInfo;
-  const diff0X = page0X - centerX;
-  const diff1X = page1X - centerX;
-  const diff0Y = page0Y - centerY;
-  const diff1Y = page1Y - centerY;
-  const dotProduct = diff0X * diff1X + diff0Y * diff1Y;
-  if (dotProduct >= 0) {
-    // The two touches go in almost the same direction.
-    return;
+  _touchInfo.touch0X = page0X;
+  _touchInfo.touch0Y = page0Y;
+  _touchInfo.touch1X = page1X;
+  _touchInfo.touch1Y = page1Y;
+
+  let pinch = false;
+
+  if (pTouch0X === page0X && pTouch0Y === page0Y) {
+    // First touch is fixed, if the vectors are collinear then we've a pinch.
+    const v1X = pTouch1X - page0X;
+    const v1Y = pTouch1Y - page0Y;
+    const v2X = page1X - page0X;
+    const v2Y = page1Y - page0Y;
+    const det = v1X * v2Y - v1Y * v2X;
+    // 0.02 is approximatively sin(0.15deg).
+    if (Math.abs(det) > 0.02 * Math.hypot(v1X, v1Y) * Math.hypot(v2X, v2Y)) {
+      return;
+    }
+    pinch = true;
+  } else if (pTouch1X === page1X && pTouch1Y === page1Y) {
+    // Second touch is fixed, if the vectors are collinear then we've a pinch.
+    const v1X = pTouch0X - page1X;
+    const v1Y = pTouch0Y - page1Y;
+    const v2X = page0X - page1X;
+    const v2Y = page0Y - page1Y;
+    const det = v1X * v2Y - v1Y * v2X;
+    if (Math.abs(det) > 0.02 * Math.hypot(v1X, v1Y) * Math.hypot(v2X, v2Y)) {
+      return;
+    }
+    pinch = true;
+  }
+
+  if (!pinch) {
```

Comment:
Do we really need the `pinch` variable, since it seems that we could just do the following?
```suggestion
  } else {
```